### ISSUE TITLE
Extending types without @key directive

### ISSUE THREAD

**Issue by malyusha** [apollographql/federation#714]
**Package Version** - "@apollo/federation": "0.22.0"

**TL;DR**
Is it possible to extend types that doesn't need keys and federation would know where to send requested fields to in parallel (as with general Query type)?

Given a schemas of 2 services: `products` and `reviews`. When federation merges these 2 schemas, no validation errors returned and schemas are composed successfully.  The key type to look at is `ProductsMutation`, defined in **Products** and extended by **Reviews**.

**Schema of `products` service**
```graphql
extend type Query {
    products(limit: Int): [Product!]
    product(id: ID!): Product
}

extend type Mutation {
    products: ProductsMutation
}

# namespaced mutation
type ProductsMutation {
    save(data: NewProductData!): SaveProductResult @goField(forceResolver: true)
}

type SaveProductResult {
    product: Product
    error: Error
}

input NewProductData {
    name: String!
}

type Product @key(fields: "id") {
    id: ID!
    url: String!
    name: String!
}
```

**Schema of `reviews` service**
```graphql
extend type Query {
    reviews: [Review!]
}

extend type Product @key(fields: "id") {
    id: ID! @external
    reviews: [Review!]
}

type Review {
    id: ID!
    text: String!
}

# extending namespaced mutation
extend type ProductsMutation {
    addReview(data: NewReviewData!): ReviewAddedResult
}

input NewReviewData {
    text: String!
    productId: ID!
}

type ReviewAddedResult {
    header: String!
    text: String!
    actionText: String!
}
```

Given two mutations:
```graphql
mutation CreateProduct($data: NewProductData!) {
  products {
    save(data: $data) {
      product {
        name
        url
        id
      }
    }
  }
}

mutation AddReview($review:NewReviewData!) {
  products {
    addReview(data:$review) {
      text
      header
      actionText
    }
  }
}
```

`CreateProduct` is sent to **Products** service, as well as `AddReview` mutation. As **Products** service doesn't know about `addReview` mutation it fails with graphql validation error. Is it expected behaviour? If expected, then why validation does not fail? If unexpected, then, I suppose it's a bug. I thought that we can extend types that doesn't need keys and federation would know where to send requested fields to in parallel (as with general `Query` type), but obviously it's not for now.

Query plan of execution these mutations:
**CreateProduct**
```
QueryPlan {
  Fetch(service: "products") {
    {
      products {
        save(data: $data) {
          product {
            name
            url
            id
          }
        }
      }
    }
  },
}
```

**AddReview**
```
QueryPlan {
  Fetch(service: "products") {
    {
      products {
        addReview(data: $review) {
          text
          header
          actionText
        }
      }
    }
  },
}
```

**Comment by pcmanus** [comment]
Thanks @malyusha for the report.

> Is it possible to extend types that doesn't need keys and federation would know where to send requested fields to in parallel (as with general Query type)?

Currently, no, this is not possible (except for the top-level Query/Mutation types that is).

In the current federation model, only "entity" types (the one with a `@key`) can be "spread" between services. Any other type in federation is considered a "value type" and should have the same exact definition (same fields) on all services (which is another way to say that such type cannot be used in the way you describe).

In other words, it is currently a validation bug that the schema above is accepted (technically, validation does not handle `extend` on value types properly and needs to be fixed).

Which is not to say that the pattern you're attempting, of namespacing operations, is not desirable. It's useful and is easily achievable in a non-federated context so we're interested in thinking how we can allow it. But this does have implications for federation that are not trivial.
